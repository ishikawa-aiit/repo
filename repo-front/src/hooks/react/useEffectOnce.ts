import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
