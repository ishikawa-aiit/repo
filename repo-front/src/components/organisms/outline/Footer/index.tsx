import { Flex } from '@chakra-ui/react'
import { FC, memo } from 'react'

import { FooterType } from './type'

export const Footer: FC<FooterType.Props> = memo(() => {
  return (
    <Flex
      data-testid="Footer"
      data-cy="Footer"
      as="footer"
      color="neutral.600"
      w="100%"
      marginTop="auto"
      fontSize="sm"
      py={3}
      justifyContent="center"
      borderTop="1px"
      borderTopColor="neutral.200"
    >
      &copy; Advanced Institute of Industrial Technology, Oikawa Project Team 2022
    </Flex>
  )
})
