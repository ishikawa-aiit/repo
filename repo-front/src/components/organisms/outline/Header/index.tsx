import { Box, Flex, Link, Text } from '@chakra-ui/react'
import Image from 'next/image'
import NextLink from 'next/link'
import { FC, memo } from 'react'
import { useRecoilState } from 'recoil'

import { HeaderType } from './type'

import { ProductConst } from '@/constants'
import { user as userState } from '@/contexts/global/atoms/User'

export const Header: FC<HeaderType.Props> = memo(() => {
  const [user] = useRecoilState(userState)

  return (
    <Flex
      data-testid="Header"
      data-cy="Header"
      as="header"
      position="sticky"
      top="0"
      left="0"
      background="white"
      w="100%"
      px={5}
      boxShadow="sm"
      // メインコンテンツで z-index を使用できるように少し大きい値にしている
      // cf: モーダルの z-index は 1400 なのでそれ未満である必要がある
      zIndex={10}
      justifyContent="space-between"
    >
      <NextLink href="/weekly-reports/create" passHref>
        <Link>
          <Box pt={2} pb={1} transitionDuration=".25s" cursor="pointer" _hover={{ opacity: '.75' }}>
            <Image
              src="/images/common/logo_repo_01.svg"
              width="110"
              height="50"
              alt={`${ProductConst.PRODUCT_CATEGORY} ${ProductConst.PRODUCT_NAME}`}
            />
          </Box>
        </Link>
      </NextLink>
      {user && (
        <Flex flexDirection="column" justifyContent="center">
          <Text data-testid="HeaderUserName">
            {user.name}
            <Text as="small" fontSize="80%" paddingLeft=".8rem">
              さん
            </Text>
          </Text>
        </Flex>
      )}
    </Flex>
  )
})
