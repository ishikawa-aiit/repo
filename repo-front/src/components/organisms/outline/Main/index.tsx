import { Box } from '@chakra-ui/react'
import { FC, memo } from 'react'

import { MainType } from './type'

export const Main: FC<MainType.Props> = memo(({ children, ...otherProps }) => {
  return (
    <Box data-testid="Main" data-cy="Main" as="main" w="100%" h="100%" overflow="hidden" {...otherProps}>
      {children}
    </Box>
  )
})
