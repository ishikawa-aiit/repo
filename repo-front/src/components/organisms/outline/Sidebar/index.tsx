import { Box, Flex, Spacer } from '@chakra-ui/react'
import { useRouter } from 'next/router'
import { FC, memo } from 'react'

import { SidebarMenuItem } from './SidebarMenuItem'
import { menuItems, bottomMenuItems } from './constants'

export const Sidebar: FC = memo(() => {
  const router = useRouter()

  return (
    <Flex
      data-testid="Sidebar"
      data-cy="Sidebar"
      as="aside"
      position="sticky"
      flexDirection="column"
      bg="primary.500"
      w="240px"
      top="68px"
      h="calc(100vh - 68px)"
    >
      <Box>
        {menuItems.map((menuItem, i) => {
          const { link } = menuItem
          const isCurrent = router.pathname === link

          return <SidebarMenuItem key={i} isCurrent={isCurrent} {...menuItem} />
        })}
      </Box>
      <Spacer />
      <Box>
        {bottomMenuItems.map((menuItem, i) => {
          const { link } = menuItem
          const isCurrent = router.pathname === link

          return <SidebarMenuItem key={menuItems.length + i} isCurrent={isCurrent} {...menuItem} />
        })}
      </Box>
    </Flex>
  )
})
