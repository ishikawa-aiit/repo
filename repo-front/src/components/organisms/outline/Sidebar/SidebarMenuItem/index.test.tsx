import { PencilIcon } from '@heroicons/react/outline'
import { render, screen } from '@testing-library/react'
import React from 'react'

import { SidebarMenuItem } from '.'

const testMenuItem = {
  link: '/test',
  name: 'テスト',
  icon: <PencilIcon data-testid="Icon" />,
  isCurrent: false,
}

jest.mock('next/link', () => {
  return ({ children, ...rest }: { children: React.ReactElement }) => {
    return React.cloneElement(children, { ...rest })
  }
})

describe('SidebarMenuItem', () => {
  it('正常にレンダリングされている', () => {
    const { name } = testMenuItem
    const { container } = render(<SidebarMenuItem {...testMenuItem} />)

    expect(screen.getByText(name)).toBeInTheDocument()
    expect(screen.getByTestId('Icon')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('指定のパスとリンクのhref属性が一致する', () => {
    const { link } = testMenuItem
    const { container } = render(<SidebarMenuItem {...testMenuItem} />)

    expect(screen.getByRole('link')).toHaveAttribute('href', link)
    expect(container).toMatchSnapshot()
  })
})

describe('props:isCurrent', () => {
  it('trueの場合、リンクがカレントとなる', () => {
    const { container } = render(<SidebarMenuItem {...testMenuItem} isCurrent />)

    expect(screen.getByTestId('SidebarCurrentMenuItem')).toBeInTheDocument()
    expect(screen.queryByTestId('SidebarMenuItem')).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('falseの場合、リンクがカレントではない', () => {
    const { container } = render(<SidebarMenuItem {...testMenuItem} />)

    expect(screen.getByTestId('SidebarMenuItem')).toBeInTheDocument()
    expect(screen.queryByTestId('SidebarCurrentMenuItem')).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
