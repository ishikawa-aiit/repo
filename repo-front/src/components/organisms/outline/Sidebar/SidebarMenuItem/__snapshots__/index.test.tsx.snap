// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarMenuItem 指定のパスとリンクのhref属性が一致する 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.white;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-0[aria-current=page] {
  background-color: primary.600;
}

.emotion-0:hover,
.emotion-0[data-hover] {
  background: primary.400;
}

.emotion-1 {
  width: 5px;
}

<div>
  <a
    class="emotion-0"
    data-cy="SidebarMenuItem"
    data-testid="SidebarMenuItem"
    href="/test"
  >
    <div
      class="emotion-1"
    >
      <svg
        aria-hidden="true"
        data-testid="Icon"
        fill="none"
        stroke="currentColor"
        stroke-width="2"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.232 5.232l3.536 3.536m-2.036-5.036a2.5 2.5 0 113.536 3.536L6.5 21.036H3v-3.572L16.732 3.732z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </div>
    テスト
  </a>
</div>
`;

exports[`SidebarMenuItem 正常にレンダリングされている 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.white;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-0[aria-current=page] {
  background-color: primary.600;
}

.emotion-0:hover,
.emotion-0[data-hover] {
  background: primary.400;
}

.emotion-1 {
  width: 5px;
}

<div>
  <a
    class="emotion-0"
    data-cy="SidebarMenuItem"
    data-testid="SidebarMenuItem"
    href="/test"
  >
    <div
      class="emotion-1"
    >
      <svg
        aria-hidden="true"
        data-testid="Icon"
        fill="none"
        stroke="currentColor"
        stroke-width="2"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.232 5.232l3.536 3.536m-2.036-5.036a2.5 2.5 0 113.536 3.536L6.5 21.036H3v-3.572L16.732 3.732z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </div>
    テスト
  </a>
</div>
`;

exports[`props:isCurrent falseの場合、リンクがカレントではない 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.white;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-0[aria-current=page] {
  background-color: primary.600;
}

.emotion-0:hover,
.emotion-0[data-hover] {
  background: primary.400;
}

.emotion-1 {
  width: 5px;
}

<div>
  <a
    class="emotion-0"
    data-cy="SidebarMenuItem"
    data-testid="SidebarMenuItem"
    href="/test"
  >
    <div
      class="emotion-1"
    >
      <svg
        aria-hidden="true"
        data-testid="Icon"
        fill="none"
        stroke="currentColor"
        stroke-width="2"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.232 5.232l3.536 3.536m-2.036-5.036a2.5 2.5 0 113.536 3.536L6.5 21.036H3v-3.572L16.732 3.732z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </div>
    テスト
  </a>
</div>
`;

exports[`props:isCurrent trueの場合、リンクがカレントとなる 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.white;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-0[aria-current=page] {
  background-color: primary.600;
}

.emotion-1 {
  width: 5px;
}

<div>
  <a
    aria-current="page"
    class="emotion-0"
    data-cy="SidebarCurrentMenuItem"
    data-testid="SidebarCurrentMenuItem"
    href="/test"
  >
    <div
      class="emotion-1"
    >
      <svg
        aria-hidden="true"
        data-testid="Icon"
        fill="none"
        stroke="currentColor"
        stroke-width="2"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.232 5.232l3.536 3.536m-2.036-5.036a2.5 2.5 0 113.536 3.536L6.5 21.036H3v-3.572L16.732 3.732z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </div>
    テスト
  </a>
</div>
`;
